src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -2213,6 +2237,95 @@ METHOD(kernel_ipsec_t, destroy, void,
 	free(this);
 }
 
+struct tunnel_counter_key {
+	uint32_t terminator_id;
+	const char* tenant_id;
+	const char* account_id;
+	const char* vrf_name;
+};
+
+static bool tunnel_counter_key_compare(const void *key, const void *other_key) {
+    const struct tunnel_counter_key *tunnel_1 = (const struct tunnel_counter_key *)key;
+    const struct tunnel_counter_key *tunnel_2 = (const struct tunnel_counter_key *)other_key;
+
+	return tunnel_1->terminator_id == tunnel_2->terminator_id &&
+			streq(tunnel_1->tenant_id, tunnel_2->tenant_id) && 
+			streq(tunnel_1->account_id, tunnel_2->account_id) && 
+			streq(tunnel_1->vrf_name, tunnel_2->vrf_name);
+}
+
+static u_int tunnel_counter_key_hash(const void *key) {
+	const struct tunnel_counter_key *tunnel_key = (const struct tunnel_counter_key *)key;
+	return chunk_hash_inc(chunk_from_thing(tunnel_key->terminator_id),
+						  chunk_hash_inc(chunk_from_str((char*)tunnel_key->tenant_id),
+						  chunk_hash_inc(chunk_from_str((char*)tunnel_key->account_id),
+						  chunk_hash(chunk_from_str((char*)tunnel_key->vrf_name)))));
+}
+
+static bool write_terminator_tunnels_oper(private_kernel_super_ipsec_t* this, struct pm_terminator_conn* const terminator_conn) {
+	const struct pm_proto_addr lookup_addr = *picomsg_terminator_conn_get_lookup_addr(terminator_conn);
+	enumerator_t* enumerator = NULL;
+	ike_sa_t* ike_sa;
+	uint64_t term_tunnels_number = 0;
+	hashtable_t *counter_map = NULL;
+	bool res = true;
+	uint32_t term_id = picomsg_terminator_conn_get_id(terminator_conn);
+
+	counter_map = hashtable_create(tunnel_counter_key_hash, tunnel_counter_key_compare, 10);
+
+	enumerator = charon->controller->create_ike_sa_enumerator(charon->controller, true);
+	while (enumerator->enumerate(enumerator, &ike_sa)) {
+		struct pm_proto_addr ike_sa_addr;
+		host_addr_to_picomsg_addr(ike_sa->get_my_host(ike_sa), &ike_sa_addr);
+		if (! pm_proto_addr_equals(&ike_sa_addr, &lookup_addr)) {
+			continue;
+		}
+		struct super_cert_info cert_info;
+		if (! parse_peer_certificate(ike_sa, &cert_info, this->peer_cert_info_local)) {
+			continue;
+		}
+
+		const struct tunnel_counter_key curr_key = {
+		    .terminator_id = term_id,
+		    .tenant_id = cert_info.tenant_id,
+		    .account_id = cert_info.account_id,
+		    .vrf_name = cert_info.vrf_name,
+		};
+
+		void *curr_val = counter_map->get(counter_map, &curr_key);
+		counter_map->put(counter_map, &curr_key, (char*)curr_val + 1);
```

Comment:
hashtable_t stores the key pointer directly. Therefore after each iteration the old pointer will become point to garbage.

In practice what most likely will happen is that the first iteration will put a pointer to 'curr_key' with the matching value 1. At the next iteration even if the new 'curr_key' changed, the 'key' pointer in the hashtable will point to the address of the **current** `curr_key` (since different iterations will most likely result in same stack layout). Therefore the key comparison function will think the old 'key' and the current 'curr_key' are the same (since they point to the same address), and so the matching value will be increment to 2.
Afterwards when this loop is existed and the hashtable is iterated, the `curr_key` pointer will point to garbage value (the pointer in the stack for the `curr_key` in the first loop, and that memory area will get overwritten by other stuff).

So instead the keys in the hashtable must be kept in the **heap**. Make sure to:
1. If `counter_map->get()` returns NULL then keep the current `curr_key` in the hashtable and set the value to 1.
2. If `counter_map->get()` returns non-NULL, then **free the old key** and `put()` with the new key (since the hashtable overwrites the key)
3. When destroying the hashtable, call `->destroy_function()` which gets a pointer to a function that suppose to free the key and value, which will suppose to `free()` the key.

Reply:
And on that same thing, the `cert_info.{tenant_id, account_id, vrf_name}` are stored in `tunnel_counter_key` as a **pointer**, but the actual strings themselves are stored in `super_cert_info` directly as a buffer. Therefore after each iteration `cert_info`'s contents will become garbage and thus the pointers in `tunnel_counter_key` will point to garbage.

Instead need to store in `tunnel_counter_key` buffers instead of pointers, and when initializing `curr_key` to copy the strings instead of just the pointers